### PR TITLE
[OpenCV] fix compatibility with VTK9

### PR DESCRIPTION
**Pull request by cenit**
After VTK9 update in #12149, VTK feature in OpenCV was broken. This patch will fix it.

### CODE REVIEW

File: ports/opencv2/CONTROL [hunk 1]
```diff
@@ -45,9 +45,5 @@ Feature: tiff
 Build-Depends: tiff
 Description: TIFF support for opencv
 
-Feature: vtk
-Build-Depends: vtk
-Description: vtk support for opencv
```
Comment (Neumann-A) [inline ports/opencv2/CONTROL:50]: hmm theoretically features cannot be removed due to `vcpkg upgrade` but having a fatal error in the portfile would also stop the upgrade. So hmm.... there is probably no good solution

File: ports/opencv4/portfile.cmake [hunk 2]
```diff
@@ -19,6 +19,7 @@ vcpkg_from_github(
       0002-install-options.patch
       0003-force-package-requirements.patch
       0004-fix-policy-CMP0057.patch
+      0005-fix-vtk9.patch
```
Comment (Neumann-A) [inline ports/opencv4/portfile.cmake:22]: Are these the upstream patches?

File: ports/opencv2/CONTROL (hunk 1 above)
Comment (cenit) [reply ports/opencv2/CONTROL:50]: the vtk feature in opencv2 was a typo. In fact as you can better inspect, there is nothing passed to vcpkg_configure_cmake
It's just a cleanup I recognized while fixing the vtk feature in opencv2/3/4

File: ports/opencv4/portfile.cmake (hunk 2 above)
Comment (cenit) [reply ports/opencv4/portfile.cmake]: more or less, I adapted fixes in OpenCV4.4 and backported them to OpenCV 4.3 and OpenCV 3.4.10

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
LGTM.

**Comment by cenit** [comment]
back to draft: there are some problems in downstream usage. Might require deeper fixes unfortunately

**Comment by JackBoosY** [comment]
@cenit Please ping me if this PR is ready for review.

**Comment by cenit** [comment]
@JackBoosY it should be better now
Can you please help me testing the configurations?
You should build `opencv3[vtk]` and `opencv4[vtk]`, and check in `config-$(triplet)-out.log` that in the line "To be built:" there is the "viz" module

```
--     To be built:                 ...... viz ......
```

It might even be possible now that static configurations are working with vtk, too. There is a section in the portfile that disables vtk for static configs, we can try to disable it and see if viz can be built

```
if (VCPKG_LIBRARY_LINKAGE STREQUAL static)
...
  if (WITH_VTK)
    message(WARNING "VTK is currently unsupported in this build configuration, turning it off")
    set(WITH_VTK OFF)
  endif()
...
endif()
```
(commenting the inner section should be enough)

**Comment by cenit** [comment]
I am testing it on some computers here, I will let you know results, any help might be very appreciated

**Comment by Neumann-A** [comment]
@cenit: any application using the viz module which can be put into `scripts\test_ports` ?

**Comment by cenit** [comment]
I can look for one. The one I am working on right now is not open source :)

**Comment by cenit** [comment]
I had to patch vtk port, it was failing otherwise in downstream projects using the library...

**Comment by cenit** [comment]
A simple test project would have detected that vtk had green mark from CI but was not working in reality, since the linking stage was failing

**Comment by cenit** [comment]
@JackBoosY PR might be ready. I have enabled VTK also on static builds. Initial tests look fine, both on OpenCV3 and on OpenCV4.

**Comment by Neumann-A** [comment]
>A simple test project would have detected that vtk had green mark from CI but was not working in reality, since the linking stage was failing

WTF? ParaView is not enough for you? It is also building a runable executable... Furthermore #11148 and #9960 were reviewed by one of the VTK maintainers. 
So if there are problems with VTK please do the following:
1. make sure it is not vcpkg related
2. make sure it is not a case of wrong usage.
3. search for upstream issues in https://gitlab.kitware.com/groups/vtk/-/issues
- if there is none with a similar problem please open up an upstream issue and mention it in the portfile.
- if there is one then there is probably already a patch. If not mention that you also ran into the problem in the upstream issue and patch it in vcpkg accordingly. Also mention the issue in the portfile.

File: ports/vtk/portfile.cmake [hunk 3]
```diff
@@ -313,3 +309,11 @@ else()
     debug_message("FILE:${CURRENT_PACKAGES_DIR}/share/vtk/${name}.cmake does not exist! No prefix correction!")
 endif()
 endforeach()
+
+file(READ ${CURRENT_PACKAGES_DIR}/share/vtk/VTK-vtk-module-find-packages.cmake VTK_FIND_MODULES)
+string(REPLACE "pugixml_vtk_module_mangle" "pugixml" VTK_FIND_MODULES "${VTK_FIND_MODULES}")
+file(WRITE ${CURRENT_PACKAGES_DIR}/share/vtk/VTK-vtk-module-find-packages.cmake "${VTK_FIND_MODULES}")
```
Comment (Neumann-A) [inline ports/vtk/portfile.cmake:315]: Do not merge until this error/change has been discussed in an upstream issue and is mentioned here. VTK team typically answers in around 48h

**Comment by cenit** [comment]
very easy check: try building this sample code without that fix
Might be that Paraview is silently patching this failure? Or that maybe if you only built a dynamically linked .exe it was failing if tried to run?

```
#include <vtkSmartPointer.h>
#include <vtkTransform.h>
#include <vtkMath.h>

int main()
{
  vtkSmartPointer<vtkTransform> transform = vtkSmartPointer<vtkTransform>::New();
  return 0;
}
```

it was broken on linux (at least) without that change :)

ps: I also suggest you to moderate your tones. Also, are you sure you have any power to veto or delay a PR so strongly?

**Comment by Neumann-A** [comment]
>Might be that Paraview is silently patching this failure? Or that maybe if you only built a dynamically linked .exe it was failing if tried

ParaView is not silently patching anything of VTK and is building on osx, linux and windows (x64&static; x86 is not tested due to a cascade failure) without issues (see vcpkg CI logs, basically everything rebuilding qt also rebuilds paraview). If it would have been patched by ParaView it would have made it directly into VTK since both ports are highly connected sharing devs. 

> ps: I also suggest you to moderate your tones. 

The tone was more disbelief than anger if you read it with that tone. 

>Also, are you sure you have any power to veto or delay a PR so strongly?

I don't have any power I can just give strong suggestions ;) But you are neither disclosing a **full** repro including the CMakeLists.txt and feature set of the installed VTK nor have you opened up an upstream issue until now.  
VTK is very active in development so delaying the PR until the root cause for the failure has been discussed or the failure has been identified as a bug and not a usage error seems reasonable to me. 
I am just considering probabilities here. From my point of view a usage error on your side has a higher probability than an error in VTK. As such I want the issue fully discussed before any action is taken and I don't mind being wrong in the end.

**Comment by Neumann-A** [comment]
CMakeLists.txt
```
cmake_minimum_required(VERSION 3.17)
project(vtk_test)
find_package(VTK CONFIG REQUIRED)
add_executable(vtk_test main.cpp)
target_link_libraries(vtk_test VTK::CommonCore VTK::CommonTransforms VTK::CommonMath)
```
main.cpp
```
#include <vtkSmartPointer.h>
#include <vtkTransform.h>
#include <vtkMath.h>

int main()
{
  vtkSmartPointer<vtkTransform> transform = vtkSmartPointer<vtkTransform>::New();
  return 0;
}
```
Terminal Output:
```
neumann@Iluinrandir:~/vtk_test$ make
Scanning dependencies of target vtk_test
[ 50%] Building CXX object CMakeFiles/vtk_test.dir/main.cpp.o
[100%] Linking CXX executable vtk_test
[100%] Built target vtk_test
```

I don't see any issues..... so somebody has to dig deeper. 
(Installed VTK: `vtk[core,opengl,paraview,qt,vtkm]:x64-linux`)

**Comment by JackBoosY** [comment]
If this PR-related issue is related to vtk, do we need to call vtk official review this PR?

**Comment by cenit** [comment]
Sorry I am on vacation from today for few days
What if you replace the find package call with
`find_package(VTK 9 NAMES vtk COMPONENTS FiltersExtraction FiltersSources FiltersTexture IOExport IOGeometry IOPLY InteractionStyle RenderingCore RenderingLOD RenderingOpenGL2 NO_MODULE REQUIRED)`
and then the target creation with
```

add_executable(vtk_test vtk_test.cpp)
target_include_directories(vtk_test PRIVATE ${VTK_INCLUDE_DIRS})
target_link_libraries(vtk_test PRIVATE ${VTK_LIBRARIES})
```
?

This config was failing for me without that fix. This is how vtk is used in the opencv community, and it was working before vtk9

**Comment by Neumann-A** [comment]
Opened up https://gitlab.kitware.com/vtk/vtk/-/issues/17977

File: ports/vtk/portfile.cmake (hunk 3 above)
Comment (cenit) [reply ports/vtk/portfile.cmake:315]: anyway, as you can see, no progress upstream and vtk is still broken here...

Comment (Neumann-A) [reply ports/vtk/portfile.cmake:315]: Just add a comment referencing the upstream issue. This way vcpkg can keep track of this patch and remove it if the upstream issue is resolved. Upstream probably also has vacation.

**Comment by cenit** [comment]
Absolutely no love upstream, despite what was vehemently said, but the problem is very real, again despite what was initially said against this PR.
The PR seems not producing any regression, while it's also fixing OpenCV's VTK feature. Can we consider it for merge?

File: ports/vtk/portfile.cmake (hunk 3 above)
Comment (Neumann-A) [inline ports/vtk/portfile.cmake:315]:
```suggestion
# see upstream issue https://gitlab.kitware.com/vtk/vtk/-/issues/17977
file(READ ${CURRENT_PACKAGES_DIR}/share/vtk/VTK-vtk-module-find-packages.cmake VTK_FIND_MODULES)
string(REPLACE "pugixml_vtk_module_mangle" "pugixml" VTK_FIND_MODULES "${VTK_FIND_MODULES}")
file(WRITE ${CURRENT_PACKAGES_DIR}/share/vtk/VTK-vtk-module-find-packages.cmake "${VTK_FIND_MODULES}")
```

**Comment by Neumann-A** [comment]
> Absolutely no love upstream

sometimes you need to poke people to get an answer ;)

But I am ok with it being merged if you add the comment about the upstream issue. 
If it is resolved the next person knows that the code may be removed/is unnecessary.
The main objection here was missing documentation why the change is necessary and how to reproduce the issue.

**Comment by cenit** [comment]
Ok, the note was added.
I also did something similar to paraview also for OpenCV in this PR, in order to keep openCV verified also for optional features.

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
Will test features after pipeline test pass.

**Comment by cenit** [comment]
> The main objection here was missing documentation why the change is necessary and how to reproduce the issue.    

Ok, but I think the title said it all. VTK feature in OpenCV is broken. No need to provide another smoking gun evidence of the problem, which was anyway provided later on, on your request

**Comment by Neumann-A** [comment]
@cenit: 
Got a patch from upstream mind trying it out in CI?
```
diff --git a/CMake/vtkModule.cmake b/CMake/vtkModule.cmake
index 4093d2d826..3d844dd50f 100644
--- a/CMake/vtkModule.cmake
+++ b/CMake/vtkModule.cmake
@@ -4409,6 +4409,15 @@ endif ()\n\n")
   foreach (_vtk_export_module IN LISTS _vtk_export_MODULES)
     get_property(_vtk_export_target_name GLOBAL
       PROPERTY "_vtk_module_${_vtk_export_module}_target_name")
+    # Use the export name of the target if it has one set.
+    get_property(_vtk_export_target_has_export_name
+      TARGET    "${_vtk_export_target_name}"
+      PROPERTY  EXPORT_NAME SET)
+    if (_vtk_export_target_has_export_name)
+      get_property(_vtk_export_target_name
+        TARGET    "${_vtk_export_target_name}"
+        PROPERTY  EXPORT_NAME)
+    endif ()
     set(_vtk_export_base "_vtk_module_find_package_${_vtk_export_module}")
     get_property(_vtk_export_packages GLOBAL
       PROPERTY "${_vtk_export_base}")
```

**Comment by JackBoosY** [comment]
json5-parser regressions will be fixed by PR #13289.

**Comment by JackBoosY** [comment]
```
CMake Error at /agent/_work/1/s/scripts/buildsystems/vcpkg.cmake:378 (_add_library):
  Target "opencv_text" links to target "LibXml2::LibXml2" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVUtils.cmake:1508 (add_library)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVModule.cmake:929 (ocv_add_library)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVModule.cmake:850 (_ocv_create_module)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVModule.cmake:1071 (ocv_create_module)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-1c7c3cc613.clean/modules/text/CMakeLists.txt:7 (ocv_define_module)


CMake Error at /agent/_work/1/s/scripts/buildsystems/vcpkg.cmake:378 (_add_library):
  Target "opencv_text" links to target "OpenSSL::Crypto" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVUtils.cmake:1508 (add_library)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVModule.cmake:929 (ocv_add_library)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVModule.cmake:850 (_ocv_create_module)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVModule.cmake:1071 (ocv_create_module)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-1c7c3cc613.clean/modules/text/CMakeLists.txt:7 (ocv_define_module)
...
```

**Comment by cenit** [comment]
@JackBoosY thanks for the help.
I should have fixed regressions

**Comment by JackBoosY** [comment]
Sadly, you didn't.

**Comment by cenit** [comment]
@JackBoosY I tried to analyse the situation.
Technically this is a regression which is already in master and it does not depend on OpenCV. There are some dependencies which are not exporting their own transitive dependencies when used and so it makes downstream projects unable to use them.
They just appeared because in this PR I expanded CI coverage for OpenCV, otherwise it gets slowly "damaged" over time by other port updates. Unfortunately I arrived late.
These regressions will take time for me to identify and fix
First of all, I need to identify which ports are depending on LibXML2 and OpenSSL and are not exporting it in their own dependencies for downstream projects. Afterwards I will have to find a way to fix them.
At first sight, they might be qt5-base (openssl) and vtk (libxml2), but I might be wrong.

**Comment by cenit** [comment]
@JackBoosY I tried to analyse the situation.
Technically this is a regression which is already in master and it does not depend on OpenCV. There are some dependencies which are not fully exporting their own transitive dependencies when used and so it makes downstream projects unable to use them.
They just appeared because in this PR I expanded CI coverage for OpenCV, otherwise it gets slowly "damaged" over time by other port updates. Unfortunately I arrived late.
These regressions will take time for me to identify and fix
First of all, I need to identify which ports are depending on LibXML2 and OpenSSL and are exporting them without a find_dependency call first to define targets. Afterwards I will have to find a way to fix them.
At first sight, they might be qt5-base (openssl) and vtk (libxml2), but I might be wrong.

**Comment by Neumann-A** [comment]
@cenit: qt5-base does not link any cmake targets. It is extracting its complete link interface from the *.prl files qmake generates. You might want to run `vcpkg_configure_cmake` with `--trace` in CI to get better logs

File: ports/opencv4/portfile.cmake [hunk 4]
```diff
@@ -401,6 +401,8 @@ set(CMAKE_AUTOMOC ON)
 set(CMAKE_AUTORCC ON)
 set(CMAKE_AUTOUIC ON)
 find_package(Qt5 COMPONENTS OpenGL Concurrent Test QUIET)
+find_package(LibXML2 QUIET)
```
Comment (Neumann-A) [inline ports/opencv4/portfile.cmake:404]: Casing wrong: The module is named `FindLibXml2` https://cmake.org/cmake/help/v3.18/module/FindLibXml2.html

Comment (cenit) [reply ports/opencv4/portfile.cmake:404]: The commit is completely wrong, also because it is downstream to opencv, while the problem is building opencv

**Comment by cenit** [comment]
@JackBoosY 
I created a new env especially for this problem.
I cannot reproduce the problem locally. Can you?
`opencv4[contrib,core,dnn,jpeg,opengl,png,tiff,vtk,webp]:x64-linux` installs without problems

**Comment by cenit** [comment]
@JackBoosY I cannot reproduce any failure locally even with the OpenCV fully featured.
Can you?

**Comment by Neumann-A** [comment]
@cenit: Probably needs a feature check of dependent ports and see if something is build with a different feature-set than you are testing with. Maybe post a` vcpkg list` result for `x64-linux` and compare against CI

**Comment by JackBoosY** [comment]
Can't repro this regression on my ubuntu machine.

**Comment by cenit** [comment]
It is required now in CI due to vcpkg-ci-opencv (which is a fully featured opencv). Still, I cannot reproduce it even with a fully featured one. We should investigate it more

**Comment by JackBoosY** [comment]
@cenit Testing all features now...

**Comment by JackBoosY** [comment]
No...I'm always stuck at the step of building llvm:
```
collect2: fatal error: ld terminated with signal 9 [Killed]
compilation terminated.
ninja: build stopped: subcommand failed.
```

**Comment by cenit** [comment]
on linux? is it the halide feature? All the other features are ok?

**Comment by JackBoosY** [comment]
```
CMake Error at /agent/_work/1/s/scripts/buildsystems/vcpkg.cmake:410 (_add_library):
  Target "opencv_text" links to target "LibXml2::LibXml2" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVUtils.cmake:1508 (add_library)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVModule.cmake:929 (ocv_add_library)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVModule.cmake:850 (_ocv_create_module)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-b809df3922.clean/cmake/OpenCVModule.cmake:1071 (ocv_create_module)
  /mnt/vcpkg-ci/buildtrees/opencv4/src/4.3.0-1c7c3cc613.clean/modules/text/CMakeLists.txt:7 (ocv_define_module)
```
I can't find _modules/text_ on my machine, which feature triggered it? And that feature should be depend on `libxml2`.

**Comment by cenit** [comment]
It should be one of the modules of opencv[contrib] but contrib does not depend on libxml2 (and adding it does not solve the problem, that's not the problem). It might be some of its dependencies that are transitively not exporting correctly the symbol downstream.
But the main problem is that I cannot reproduce it locally. Can you?

**Comment by JackBoosY** [comment]
@cenit No, I can't. I passed the configure step.

**Comment by cenit** [comment]
And if you add all opencv features (apart from halide, which triggers the llvm problem you had before)?

**Comment by JackBoosY** [comment]
@cenit Only feature halide triggered llvm.

**Comment by cenit** [comment]
I mean, can you pass configure step with all features enabled (apart from halide)? I can, and so I cannot reproduce CI results locally

**Comment by Neumann-A** [comment]
please c&p the output of `vcpkg.exe depend-info opencv[*]` here. I can take a look and compare against CI

**Comment by JackBoosY** [comment]
```
root@root:/home/work/12785/vcpkg# ./vcpkg  install opencv4[*]
Computing installation plan...
The following packages will be built and installed:
  * halide[core]:x64-linux
  * llvm[clang,core,disable-abi-breaking-checks,disable-assertions,disable-clang-static-analyzer,enable-rtti,lld,tools]:x64-linux
  * openblas[core]:x64-linux
    opencv4[ade,contrib,core,cuda,dnn,eigen,ffmpeg,gdcm,halide,ipp,jasper,jpeg,nonfree,openexr,opengl,openmp,ovis,png,qt,sfm,tbb,tiff,vtk,webp,world]:x64-linux
  * pegtl-2[core]:x64-linux
  * proj4[core,database]:x64-linux
  * qt5-base[core]:x64-linux
  * tbb[core]:x64-linux
  * utfcpp[core]:x64-linux
  * vtk[core]:x64-linux
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-linux...
^C
```

**Comment by cenit** [comment]
> vcpkg.exe depend-info opencv4[*]

```
./vcpkg depend-info opencv[nonfree,ade,contrib,dnn,eigen,ffmpeg,gdcm,ipp,jasper,jpeg,openexr,opengl,openmp,png,qt,sfm,tiff,vtk]
zlib:
openssl-unix:
libjpeg-turbo:
liblzma:
openssl: openssl-unix
libpng: zlib
szip:
egl-registry:
brotli:
bzip2:
tiff: libjpeg-turbo, liblzma, zlib
opengl-registry: egl-registry
libffi:
xxhash:
libogg:
hdf5[szip, zlib]: szip, zlib
curl[openssl, ssl, non-http]: openssl, zlib
libiconv:
giflib:
gflags:
libwebp[simd, nearlossless]:
freetype[png, bzip2]: brotli, bzip2, libpng, zlib
ragel:
sqlite3[tool]:
dirent:
expat:
zstd:
lz4: xxhash
opengl:
netcdf-c: curl, hdf5
openjpeg:
pcre2:
pegtl-2:
proj4[database]: sqlite3
pugixml:
python3: libffi, openssl, zlib
utfcpp:
libtheora: libogg
fontconfig: dirent, expat, freetype, libiconv
freeglut:
eigen3:
glew:
double-conversion:
glog: gflags
harfbuzz: freetype, ragel
icu:
jsoncpp:
leptonica: giflib, libjpeg-turbo, libpng, libwebp, tiff, zlib
libharu[notiffsymbols]: libpng, tiff, zlib
libpq[zlib, openssl]: openssl, zlib
angle: egl-registry, opengl-registry
libxml2: libiconv, liblzma, zlib
vtk: double-conversion, eigen3, expat, freetype, glew, hdf5, jsoncpp, libharu, libjpeg-turbo, liblzma, libogg, libpng, libtheora, libxml2, lz4, netcdf-c, pegtl-2, proj4, pugixml, sqlite3, tiff, utfcpp, zlib
ceres: eigen3, glog
tesseract: leptonica
ade:
ffmpeg[swscale, swresample, postproc, avresample, avformat, avfilter, avdevice, avcodec]:
qt5-base: angle, double-conversion, egl-registry, fontconfig, freetype, harfbuzz, icu, libjpeg-turbo, libpng, libpq, openssl, pcre2, sqlite3, zlib, zstd
gdcm: expat, openjpeg, zlib
protobuf:
jasper: freeglut, libjpeg-turbo, opengl
openexr: python3, zlib
opencv4[tiff, sfm, qt, webp, png, openmp, ipp, gdcm, jpeg, jasper, ffmpeg, eigen, vtk, openexr, dnn, opengl, nonfree, contrib, ade]: ade, ceres, eigen3, ffmpeg, gdcm, gflags, glog, hdf5, jasper, libjpeg-turbo, libpng, libwebp, openexr, opengl, protobuf, qt5-base, tesseract, tiff, vtk, zlib
opencv[tiff, sfm, qt, webp, png, openmp, ipp, gdcm, jpeg, jasper, ffmpeg, eigen, vtk, openexr, dnn, opengl, nonfree, contrib, ade]: opencv4
```

**Comment by cenit** [comment]
I can build it locally on ubuntu 20.04 without any problem

**Comment by Neumann-A** [comment]
@cenit Have you tried to built libxml2 first and after that build everything opencv4 needs? I already checked the depend-info vs CI and it looks like some port has a hidden dependency on libxml2.

**Comment by cenit** [comment]
Good idea. I will do it tonight and check if I can recover the error.

**Comment by cenit** [comment]
error disappeared from Linux and appeared in macOS. Due to the similarities in the build toolchains, it is a hidden dependency somewhere that just was casually removed from the Linux baseline and added to the osx recently

**Comment by cenit** [comment]
error disappeared from Linux and appeared in macOS. Due to the similarities in the build toolchains, it looks really like it is due to hidden dependencies on OpenSSL and LibXML2 somewhere in some ports that just was casually removed from the Linux baseline and added to the osx recently

**Comment by JackBoosY** [comment]
@cenit I think so. That's affect the *inx OS.

**Comment by cenit** [comment]
> @cenit Have you tried to built libxml2 first and after that build everything opencv4 needs? I already checked the depend-info vs CI and it looks like some port has a hidden dependency on libxml2.

@Neumann-A I did it. And with much sadness, after having created an empty environment, installed libxml2 manually at the beginning and then reinstalled everything, it built successfully.

**Comment by Neumann-A** [comment]
@cenit: add `--trace-expand` as an option to `vcpkg_configure_cmake` in opencv4 and get the full log. With this the logs can be searched for the targets and this should reveal where the targets are linked into.

**Comment by cenit** [comment]
You mean in CI? Because to do it locally i'd have to be able to reproduce the error here, first

**Comment by Neumann-A** [comment]
@cenit:
```
	Line 93723: /Users/vagrant/Data/installed/x64-osx/share/vtk/VTK-targets.cmake(745):  set_target_properties(VTK::libxml2 PROPERTIES INTERFACE_LINK_LIBRARIES LibXml2::LibXml2 )
	Line 131947: /Users/vagrant/Data/installed/x64-osx/share/tesseract/TesseractTargets.cmake(55):  set_target_properties(libtesseract PROPERTIES INTERFACE_LINK_LIBRARIES $<LINK_ONLY:pthread>;leptonica;ZLIB::ZLIB;BZip2::BZip2;LibXml2::LibXml2;lz4::lz4;LibLZMA::LibLZMA;$<$<NOT:$<CONFIG:DEBUG>>:/Users/vagrant/Data/installed/x64-osx/lib/liblzo2.a>;$<$<CONFIG:DEBUG>:/Users/vagrant/Data/installed/x64-osx/debug/lib/liblzo2.a>;$<$<NOT:$<CONFIG:DEBUG>>:/Users/vagrant/Data/installed/x64-osx/lib/libzstd.a>;$<$<CONFIG:DEBUG>:/Users/vagrant/Data/installed/x64-osx/debug/lib/libzstdd.a>;OpenSSL::Crypto )
```
Seems like an issue with Tesseract? The Tesseract config seems to not have any `find_dependency` calls if I read the log correctly. 
```
/Users/vagrant/Data/installed/x64-osx/share/tesseract/TesseractConfig.cmake(22):  include(/Users/vagrant/Data/installed/x64-osx/share/tesseract/TesseractTargets.cmake )
/Users/vagrant/Data/installed/x64-osx/share/tesseract/TesseractConfig.cmake(24):  find_package(Leptonica REQUIRED )
/Users/vagrant/Data/installed/x64-osx/share/tesseract/TesseractConfig.cmake(30):  SET(Tesseract_VERSION 4.1.1 )
/Users/vagrant/Data/installed/x64-osx/share/tesseract/TesseractConfig.cmake(31):  SET(Tesseract_VERSION_MAJOR 4 )
/Users/vagrant/Data/installed/x64-osx/share/tesseract/TesseractConfig.cmake(32):  SET(Tesseract_VERSION_MINOR 1 )
/Users/vagrant/Data/installed/x64-osx/share/tesseract/TesseractConfig.cmake(39):  set(Tesseract_INCLUDE_DIRS /Users/vagrant/Data/installed/x64-osx/share/tesseract/../../include;/Users/vagrant/Data/installed/x64-osx/share/tesseract/../../include/tesseract )
/Users/vagrant/Data/installed/x64-osx/share/tesseract/TesseractConfig.cmake(45):  set(Tesseract_LIBRARIES libtesseract )
```

**Comment by cenit** [comment]
Good catch!

**Comment by cenit** [comment]
ok tesseract can depend also on libarchive, which depends on openssl and libxml2. It's not mandatory, somehow was found on macOS and so incorporated.
Found the link, finally. I will add libarchive explicitly for all, and then fix the symbols going downstream.
Thanks for the hint @Neumann-A

**Comment by JackBoosY** [comment]
Only osx regression left now.

**Comment by JackBoosY** [comment]
Pinging @cenit for response. Is work still being done for this PR?

**Comment by cenit** [comment]
the issue is still present on master and this PR fixes it.
Unfortunately it uncovered also problems, which are present in master! I don't have much time to dig them, I hope during the weekend to be able to catch-up

**Comment by cenit** [comment]
If you merge this PR, you won't do any damage to current master situation, on the opposite you would fix a problem. You would also highlight (finally) that some port updates broke some opencv features!

**Comment by JackBoosY** [comment]
But `opencv4:x64-osx` still fails.

**Comment by cenit** [comment]
I still have to work on it as I said.
It's not the base opencv4 port which fails, it's the opencv4 version triggered by `scripts/test_ports/vcpkg-ci-opencv` that fails. In that one many features are tested and it would be awesome to have these checks applied even before #14118 is merged

**Comment by JackBoosY** [comment]
I tend to merge it after you complete this PR, because all the CI tests must turn green : )

**Comment by cenit** [comment]
@strega-nil is pangolin regression due to this PR? I doubt it...
Please let me know if this PR can be considered for merging
I also already saw that there are problems with brand new Xcode and Ceres, which are triggering downstream problems in OpenCV. I am not going to fix them here because they are 
1) out of topic for this PR (in any case it's just a 
```
set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
missing at the end of CeresConfig.cmake)
2) still not triggered by CI (maybe it's not updated to latest Xcode)

**Comment by strega-nil** [comment]
@BillyONeal is the new person on duty, so let's see what he has to say.

**Comment by cenit** [comment]
after a rerun pangolin disappeared and opencv4 on macOS returned. I am unable to reproduce it locally on my machine.
Because of the regression, I am almost sure that the PR still cannot be accepted

**Comment by cenit** [comment]
I am trying some "blind fixes" using directly the azure pipeline here to verify them, due to the fact that locally it works.
I will try some experiments exploiting some discussions above which led to some suggestions

**Comment by BillyONeal** [comment]
Unfortunately I haven't seen a build where `pangolin` failed; it isn't one of the ones that have often failed in our CI anyway but there could be some form of races in there.

At least in the current build it looks like it doesn't like generator expressions in a place?

```
CMake Warning at cmake/OpenCVUtils.cmake:1576 (message):
  Unexpected CMake generator expression:
  $<$<NOT:$<CONFIG:DEBUG>>:/Users/vagrant/Data/installed/x64-osx/lib/libwebp.a>
Call Stack (most recent call first):
  CMakeLists.txt:1085 (ocv_get_all_libs)
```

I'm not sure where `OpenCVUtils.cmake` comes from...

**Comment by BillyONeal** [comment]
I also observe pangolin says `Supports: !uwp & !osx` so I'm not sure how it could affect an osx run..

**Comment by cenit** [comment]
> I also observe pangolin says `Supports: !uwp & !osx` so I'm not sure how it could affect an osx run..

https://github.com/microsoft/vcpkg/runs/1515708729

this is the run where pangolin failed. In fact, osx went fine there and windows failed.
I will investigate this generator expression regression in opencv and let you know

**Comment by cenit** [comment]
wait a minute: why is opencv not tested in ANY pipeline apart osx??
It is marked as a cascade failure and so not tested 😨 

This PR is arriving very late! So late that OpenCV is already broken on master?? 😢

**Comment by cenit** [comment]
@BillyONeal in any case now this PR has a green mark, I hope it can be considered for merging because it contains an opencv test which is verifying important features excluded from the default build which otherwise often regress

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
@BillyONeal Ping for review and merge this PR.

**Comment by BillyONeal** [comment]
Thanks for your help!

**Pull request merged**